src/core/markovProcessGraph.js

```diff
@@ -0,0 +1,579 @@
+// one node.
+const SEED_ADDRESS = NodeAddress.fromParts(["sourcecred", "core", "SEED"]);
+const SEED_DESCRIPTION = "\u{1f331}"; // SEEDLING
+
+// Node address prefix for epoch nodes.
+const EPOCH_PREFIX = NodeAddress.fromParts(["sourcecred", "core", "EPOCH"]);
+
+export type EpochNodeAddress = {|
+  +type: "EPOCH_NODE",
+  +owner: NodeAddressT,
+  +epochStart: TimestampMs,
+|};
+
+export function epochNodeAddressToRaw(addr: EpochNodeAddress): NodeAddressT {
+  return NodeAddress.append(
+    EPOCH_PREFIX,
+    String(addr.epochStart),
+    ...NodeAddress.toParts(addr.owner)
+  );
+}
+
+export function epochNodeAddressFromRaw(addr: NodeAddressT): EpochNodeAddress {
+  if (!NodeAddress.hasPrefix(addr, EPOCH_PREFIX)) {
+    throw new Error("Not an epoch node address: " + NodeAddress.toString(addr));
+  }
+  const epochPrefixLength = NodeAddress.toParts(EPOCH_PREFIX).length;
+  const parts = NodeAddress.toParts(addr);
+  const epochStart = +parts[epochPrefixLength];
+  const owner = NodeAddress.fromParts(parts.slice(epochPrefixLength + 1));
+  return {
+    type: "EPOCH_NODE",
+    owner,
+    epochStart,
+  };
+}
+
+// Prefixes for fibration edges.
+const FIBRATION_EDGE = EdgeAddress.fromParts([
+  "sourcecred",
+  "core",
+  "fibration",
+]);
+const EPOCH_PAYOUT = EdgeAddress.append(FIBRATION_EDGE, "EPOCH_PAYOUT");
+const EPOCH_WEBBING = EdgeAddress.append(FIBRATION_EDGE, "EPOCH_WEBBING");
+// Prefixes for seed edges.
+const SEED_IN = EdgeAddress.fromParts(["sourcecred", "core", "SEED_IN"]);
+const SEED_OUT = EdgeAddress.fromParts(["sourcecred", "core", "SEED_OUT"]);
+
+export type FibrationOptions = {|
+  // List of node prefixes for temporal fibration. A node with address
+  // `a` will be fibrated if `NodeAddress.hasPrefix(node, prefix)` for
+  // some element `prefix` of `what`.
+  +what: $ReadOnlyArray<NodeAddressT>,
+  // Transition probability for payout edges from epoch nodes to their
+  // owners.
+  +beta: TransitionProbability,
+  // Transition probability for webbing edges from an epoch node to the
+  // next epoch node for the same owner.
+  +gammaForward: TransitionProbability,
+  +gammaBackward: TransitionProbability,
+|};
+export type SeedOptions = {|
+  +alpha: TransitionProbability,
+|};
+
+const COMPAT_INFO = {type: "sourcecred/markovProcessGraph", version: "0.1.0"};
+
+export type MarkovProcessGraphJSON = Compatible<{|
+  +nodes: {|+[NodeAddressT]: MarkovNode|},
+  +edges: {|+[MarkovEdgeAddressT]: MarkovEdge|},
+  +scoringAddresses: $ReadOnlyArray<NodeAddressT>,
+|}>;
+
+export class MarkovProcessGraph {
+  _nodes: Map<NodeAddressT, MarkovNode>;
+  _edges: Map<MarkovEdgeAddressT, MarkovEdge>;
+  _scoringAddresses: Set<NodeAddressT>;
+
+  constructor(
+    nodes: Map<NodeAddressT, MarkovNode>,
+    edges: Map<MarkovEdgeAddressT, MarkovEdge>,
+    scoringAddresses: Set<NodeAddressT>
+  ) {
+    this._nodes = nodes;
+    this._edges = edges;
+    this._scoringAddresses = scoringAddresses;
```

Comment:
I really like that we take an explicit set of scoring addresses, rather than prefix matchers.